src/anti_hoisting.rs

```diff
@@ -0,0 +1,172 @@
+        
+        // invalid if output does not contain ${nickname}
+        if !output.contains(AntiHoisting::NICKNAME_PLACEHOLDER) {
+            return Err(Error::custom("`output` must contain `${nickname}`"));
+        };
+
+        //TODO validation
+
+        Ok(Configuration {
+            enabled,
+            max_nickname_length,
+            min_nickname_length,
+            trigger,
+            output: output.to_owned(),
+            ignore_roles,
+        })
+    }
+}
+
+pub struct AntiHoisting {}
+
+impl AntiHoisting {
+    pub const INTENTS: Intents = Intents::GUILD_MEMBERS;
+    pub const NICKNAME_PLACEHOLDER: &str = "${nickname}";
+
+    /// run the main anti-hoisting logic
+    pub async fn serve(config: Configuration, mut events: Receiver<Arc<Event>>, http: Arc<Client>) {
+        loop {
+            let event = events.recv().await;
+
+            // match events where a member's name is hoisted
+            let hoisted_member = match event.as_deref() {
+                Err(RecvError::Closed) => return,
+                Err(_) => continue,
+                Ok(Event::MemberUpdate(m)) if m.nick.as_ref().map_or(false, |nick| AntiHoisting::is_hoisted(&config.trigger, &nick))  => m,
+                Ok(_) => continue,
+            };
+
+            // skip member with a role that bypasses anti-hoisting
+            if hoisted_member.roles.iter().any(|role| config.ignore_roles.contains(role)) {
+                debug!("Member {:#?} has a role that bypasses anti-hoisting, skipping", hoisted_member.user.name);
+                continue;
+            };
+            
+            // Only nicknames are supported
+            let Some(ref old_nickname) = hoisted_member.nick else {
+                debug!("Member {:#?} has not changed a nickname, skipping", hoisted_member.user.name);
+                continue;
+            };
+
+            let new_nickname = AntiHoisting::transform(&old_nickname, &config.output);
+
+            // the output can produce a nickname longer than max_nickname_length
+            // for now the strategy is to not change the nickname
+            if new_nickname.len() > config.max_nickname_length as usize {
+                continue;
+            }
+
+            let result = Self::change_nickname(&hoisted_member, &new_nickname, &http).await;
+
+            let member= match result {
+                Err(err) => {
+                    // permission errors
+                    warn!("Failed to change nickname: {err}");
+                    continue;
+                },
+                Ok(m) if m.nick.as_ref().map_or(false, |nick| *nick == new_nickname) => m,
+                Ok(_) => continue,
+            };
+
+            info!("Member has tried to hoist with name: {:#?} set new nickname to: {:#?}", hoisted_member.nick, &new_nickname);
+        }
+    }
+
+    async fn change_nickname(hoisted_member: &Box<MemberUpdate>, new_nickname: &str, http: &Arc<Client>) -> anyhow::Result<Member> {
```

Comment:
```suggestion
    async fn change_nickname(hoisted_member: &MemberUpdate, new_nickname: &str, http: &Arc<Client>) -> anyhow::Result<Member> {
```
It makes more sense to use a reference in the signature here. Box<T> references to <T> anyway.